libfwupdplugin/fu-fmap-firmware.c

```diff
@@ -0,0 +1,265 @@
+}
+
+/* brute force linear search */
+static gboolean
+fmap_lsearch (const guint8 *image, gsize len, gsize *offset, GError **error)
+{
+	gsize i;
+	gboolean fmap_found = FALSE;
+
+	if (offset == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "offset return not valid");
+		return FALSE;
+	}
+
+	for (i = 0; i < len - strlen(FMAP_SIGNATURE); i++) {
+		if (!memcmp(&image[i],
+		            FMAP_SIGNATURE,
+		            strlen(FMAP_SIGNATURE))) {
+			fmap_found = TRUE;
+			break;
+		}
+	}
+
+	if (!fmap_found) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "fmap not found using linear search");
+		return FALSE;
+	}
+
+	if (i + fmap_size ((FuFmap *)&image[i]) > len) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "malformed fmap too close to end of image");
+
+		return FALSE;
+	}
+
+	*offset = i;
+	return TRUE;
+}
+
+/* if image length is a power of 2, use binary search */
+static gboolean
+fmap_bsearch (const guint8 *image, gsize len, gsize *offset, GError **error)
+{
+	gsize i;
+	gint stride;
+	gboolean fmap_found = FALSE;
+
+	if (offset == NULL) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "offset return not valid");
+		return FALSE;
+	}
+
+	/*
+	 * For efficient operation, we start with the largest stride possible
+	 * and then decrease the stride on each iteration. Also, check for a
+	 * remainder when modding the offset with the previous stride. This
+	 * makes it so that each offset is only checked once.
+	 */
+	for (stride = len / 2; stride >= 1; stride /= 2) {
+		if (fmap_found)
+			break;
+
+		for (i = 0; i < len - strlen(FMAP_SIGNATURE); i += stride) {
+			if ((i % (stride * 2) == 0) && (i != 0))
+				continue;
+			if (!memcmp (&image[i],
+				     FMAP_SIGNATURE,
+				     strlen (FMAP_SIGNATURE))) {
+				fmap_found = TRUE;
+				break;
+			}
+		}
+	}
+
+	if (!fmap_found) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "fmap not found using binary search");
+		return FALSE;
+	}
+
+	if (i + fmap_size ((FuFmap *)&image[i]) > len)
+		return FALSE;
+
+	*offset = i;
+	return TRUE;
+}
+
+static gint
+popcnt (guint u)
+{
+	gint count;
+
+	/* K&R method */
+	for (count = 0; u; count++)
+		u &= (u - 1);
+
+	return count;
+}
+
+static gboolean
+fmap_find (const guint8 *image, gsize image_len, gsize *offset, GError **error)
+{
+	gboolean ret;
+
+	if (image == NULL || image_len == 0) {
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_INVALID_DATA,
+				     "invalid image");
+		return FALSE;
+	}
+
+	if (popcnt (image_len) == 1)
+		ret = fmap_bsearch (image, image_len, offset, error);
+	else
+		ret = fmap_lsearch (image, image_len, offset, error);
+
+	if (!ret) {
+		g_prefix_error (error, "failed fmap_find: ");
+		return FALSE;
+	}
+
+	return TRUE;
```

Comment:
could we redo this in a nicer way? e.g. 

    if (popcnt (image_len) == 1) {
        if (!fmap_bsearch (image, image_len, offset, error)) {
            g_prefix_error (error, "failed fmap_find for cnt 1: ");
            return FALSE;
        }
    }

    if (!fmap_lsearch (image, image_len, offset, error)) {
        g_prefix_error (error, "failed fmap_find: ");
        return FALSE;
    }

    return TRUE;

Then the error message tells us which path was taken.

Reply:
Done.